docs: add URL variables in AsyncAPI Document (#1890)
asyncapi/website: pages/docs/concepts/asyncapi document/variable-url.md

---
title: Variable URL
weight: 275
---

> [Event-driven architecture](/docs/tutorials/getting-started/event-driven-architectures) benefits from a defined base URL and rules for URL variables. You can easily manage multiple API endpoints, handling various server configurations and environments.

Reviewer: If you want to keep this paragraph, I'd make it like this:

```suggestion
Servers benefit from a defined base URL and rules for URL variables. You can easily manage multiple endpoints, handling various server configurations and environments.
```

But honestly, I don't think this paragraph is adding any relevant information for the topic so I'd completely remove it.

> URL variables are placeholders for values you can replace during runtime. AsyncAPI enables you to construct dynamic URLs with query parameters and additional information while enhancing the flexibility and maintainability of your API specifications.

Reviewer: ```suggestion
URL variables are placeholders for values you can replace during runtime. AsyncAPI enables you to construct dynamic URLs while enhancing the flexibility and maintainability of your AsyncAPI documents.
```

Also, what do you mean with "while enhancing the flexibility and maintainability"?

## Add dynamic variables

> To add variables to the URL using AsyncAPI between two servers. You can use the `server.url` and `components.serverVariables` fields to enable reusable variables across multiple servers.

Reviewer: ```suggestion
Use the `server.url` and `server.variables` to add variables to a server URL. Leverage `components.serverVariables` to enable reusable variable definitions across multiple servers.
```

The diagram below describes how to use variable urls in AsyncAPI.

```mermaid
graph TD

    A[AsyncAPI Server]
    B[Components]
    C[Server URL]

style A fill:#47BCEE,stroke:#47BCEE;
style B fill:#47BCEE,stroke:#47BCEE;
style C fill:#47BCEE,stroke:#47BCEE;

  B -->|Defines serverVariables| C
  C -->|Reusable variables| A

> ```

Reviewer: I honestly don't think a diagram is helping here. I'd completely remove it.

### Servers section

Define the servers section in your AsyncAPI document, and include the base URLs for your API servers. Use placeholders enclosed in curly braces {} to represent the variables in the server URL. For example:

```yaml
servers:
  production:
    url: 'https://{subdomain}.example.com:{port}/v1'
    description:
      Production server:
      variables:
        domain:
          enum:
            - development
            - staging
            - production
        port:

>           default: '8080'

Reviewer: ```suggestion
    variables:
      subdomain:
        enum:
          - development
          - staging
          - production
      port:
        default: '8080'
```

Reviewer: A few things happened here:
* Indentation of `variables:` is not correct. I indented it 2 spaces back to align with `url`.
* The name of the variable is now `subdomain` instead of `domain`.
* I removed the `description` field because it's adding no value to the example.

Reviewer: To prevent future errors with the indentation, I suggest that you always test your snippets in https://studio.asyncapi.com.

```

### `serverVariables` section

> Define the components.serverVariables section in your AsyncAPI document. For each variable used in the server URLs, provide a default value and an optional description:

Reviewer: The `serverVariables` section under `components` is meant for reusability, just so you don't define the same variables over and over in the `variables:` field under the `servers` section (like in the example above). Let me give you an example following the previous one:

```yaml
servers:
  production:
    url: 'https://{subdomain}.example.com:{port}/v1'
    variables:
      subdomain:
        enum:
          - development
          - staging
          - production
      port:
        default: '8080'
  staging:
    url: 'https://{subdomain}.example.com:{port}/v1'
    variables:
      subdomain:
        enum:
          - development
          - staging
          - production
      port:
        default: '8080'
  development:
    url: 'https://{subdomain}.example.com:{port}/v1'
    variables:
      subdomain:
        enum:
          - development
          - staging
          - production
      port:
        default: '3000'
```

As you can see, the `subdomain` variable definition is repeated multiple times. If we now wanted to change the definition of `subdomain`, we would have to do it in all of them. To avoid repeating ourselves, we have `components.serverVariables`:

```yaml
servers:
  production:
    url: 'https://{subdomain}.example.com:{port}/v1'
    variables:
      subdomain:
        $ref: '#/components/serverVariables/subdomain'
      port:
        default: '8080'
  staging:
    url: 'https://{subdomain}.example.com:{port}/v1'
    variables:
      subdomain:
        $ref: '#/components/serverVariables/subdomain'
      port:
        default: '8080'
  development:
    url: 'https://{subdomain}.example.com:{port}/v1'
    variables:
      subdomain:
        $ref: '#/components/serverVariables/subdomain'
      port:
        default: '8080'

components:
  serverVariables:
    subdomain:
      enum:
        - development
        - staging
        - production
```

Now the definition is just in one place and all the occurrences of `subdomain` are just pointers (or as we call it, references) to the definition under `components.serverVariables`.

You may have noticed that the same problem happens with the `port` variable. I'll leave it up to you to extract it to the `components.serverVariables` section to avoid repetition. Make sure you open https://studio.asyncapi.com and try it there.

```yaml
components:
  serverVariables:

>     domain:

Reviewer: ```suggestion
    subdomain:
```

      default: 'api'

>       description: The domain of the API server port

Reviewer: ```suggestion
      description: The subdomain
```

>         default: '8080'
>         description: The port of the API server

Reviewer: ```suggestion
```

```

> ### Define domain and port variables

Reviewer: This is very specific and it might make people think that server variables are for defining subdomains and ports. You can but this is just one of the multiple usages.
